src/v/storage/segment_index.cc

```diff
@@ -213,14 +252,31 @@ ss::future<> segment_index::flush() {
     _needs_persistence = false;
     auto backing_file = co_await open();
     co_await backing_file.truncate(0);
-    auto out = co_await ss::make_file_output_stream(
-      ss::file(backing_file.dup()));
-    auto b = serde::to_iobuf(_state.copy());
-    for (const auto& f : b) {
-        co_await out.write(f.get(), f.size());
-    }
-    co_await out.flush();
-    co_await out.close();
+    auto out = co_await ss::make_file_output_stream(std::move(backing_file));
+
+    std::exception_ptr ex;
+    try {
+        auto b = serde::to_iobuf(_state.copy());
+        for (const auto& f : b) {
+            co_await out.write(f.get(), f.size());
+        }
+        co_await out.flush();
+    } catch (...) {
+        ex = std::current_exception();
+    }
+
+    // Close, swallowing exception if ex is already set
+    try {
+        co_await out.close();
+    } catch (...) {
+        if (!ex) {
```

Comment:
is this the sort of case where we might want to deploy a `std::nested_exception`?

Reply:
Replaced this with use of ss::with_file, which I think implicitly benefits from the nested exception handling in finally()